internal/webhooks/patch_validation.go

```diff
@@ -0,0 +1,333 @@
+	var allErrs field.ErrorList
+	if selector.MatchResources.InfrastructureCluster == nil && selector.MatchResources.ControlPlane == nil &&
+		selector.MatchResources.MachineDeploymentClass == nil && len(selector.MatchResources.MachineDeploymentClass.Names) == 0 {
+		return append(allErrs,
+			field.Invalid(
+				path,
+				selector,
+				"no selector enabled",
+			))
+	}
+	if selector.MatchResources.InfrastructureCluster != nil && class.Spec.Infrastructure.Ref != nil {
+		if selectorMatchTemplate(selector, class.Spec.Infrastructure.Ref) {
+			return nil
+		}
+	}
+	if selector.MatchResources.ControlPlane != nil && class.Spec.ControlPlane.Ref != nil {
+		if selectorMatchTemplate(selector, class.Spec.ControlPlane.Ref) {
+			return nil
+		}
+		if class.Spec.ControlPlane.MachineInfrastructure != nil {
+			if selectorMatchTemplate(selector, class.Spec.ControlPlane.MachineInfrastructure.Ref) {
+				return nil
+			}
+		}
+	}
+	if selector.MatchResources.MachineDeploymentClass != nil {
+		for _, name := range selector.MatchResources.MachineDeploymentClass.Names {
+			for _, md := range class.Spec.Workers.MachineDeployments {
+				if md.Class == name {
+					if selectorMatchTemplate(selector, md.Template.Infrastructure.Ref) {
+						return nil
+					}
+					if selectorMatchTemplate(selector, md.Template.Bootstrap.Ref) {
+						return nil
+					}
+				}
+			}
+		}
+	}
+	// if the code has not returned at this point there is no matching template in the ClusterClass. Return an error.
+	return append(allErrs,
+		field.Invalid(
+			path,
+			selector,
+			"selector did not match any template in the ClusterClass",
+		))
+}
+
+// selectorMatchTemplate returns true if APIVersion and Kind for the given selector match the reference.
+func selectorMatchTemplate(selector clusterv1.PatchSelector, reference *corev1.ObjectReference) bool {
+	return selector.Kind == reference.Kind && selector.APIVersion == reference.APIVersion
+}
+
+var validOps = sets.NewString("add", "replace", "remove")
+
+func validateJSONPatches(jsonPatches []clusterv1.JSONPatch, variables []clusterv1.ClusterClassVariable, path *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	variableSet := getClusterClassVariablesMap(variables)
+
+	for i, jsonPatch := range jsonPatches {
+		if !validOps.Has(jsonPatch.Op) {
+			allErrs = append(allErrs,
+				field.NotSupported(
+					path.Index(i).Child("op"),
+					jsonPatch,
+					validOps.List(),
+				))
+		}
+
+		if !strings.HasPrefix(jsonPatch.Path, "/spec/") {
+			allErrs = append(allErrs,
+				field.Invalid(
+					path.Index(i),
+					jsonPatch,
+					"jsonPatch path must start with \"/spec/\".",
+				))
+		}
+
+		// return early if the jsonPatch does not have "replace" or "add" op. Additional validation is not needed.
+		if jsonPatch.Op != "add" && jsonPatch.Op != "replace" {
+			return allErrs
+		}
+
+		if jsonPatch.Value == nil && jsonPatch.ValueFrom == nil {
```

Comment:
Maybe we should move the value validation in its own `validateJSONPatchValue` func which we then call here. I think it would make this func a bit easier to digest (especially things like "hey we only validate value for `add` and `replace`")